autoPyTorch/api/base_task.py

```diff
@@ -183,6 +183,9 @@ def __init__(
         self.trajectory: Optional[List] = None
         self.dataset_name: Optional[str] = None
         self.cv_models_: Dict = {}
+        self.precision: Optional[int] = None
+        self.opt_metric: Optional[str] = None
+        self.dataset: Optional[BaseDataset] = None
```

Comment:
Although each `BaseDataset` shares the same address, it is not a good idea to have `BaseDataset` everywhere (because we do not know when and where and how it can change).
So can we take `self.dataset` as an argument when we would like to use?

Reply:
so `self.dataset` is only modified in the task.search method. It is already an argument to the functions `fit` and `refit` so that allows the user to use a different dataset to refit all the models or even fitting a pipeline. and the `fit_ensemble` as it is intended to only be used after `search()` does not have an option to take a datasert as we need to ensure its the same dataset as the search was called on. Taking it as an argument does not seem to provide that much advantanges especially considering creating a dataset is hard, we dont have a function currently that will create this dataset and the feature validators and fit and transform them for the user. And needing to give a dataset object for each function would thus become tougher. I have merged a `fit_pipeline` [PR](https://github.com/automl/Auto-PyTorch/pull/173) to the cocktails branch which would introduce the create_dataset method. I'll soon make a PR to add this to `development` as well. Maybe we can revisit this conversation then?

Reply:
Thanks for the clarification:)
In fact, this is very general discussion.
It is just my opinion and it will be great if we can revisit some times soon.
FYI, it is worth taking a look at the concept "encapsulation of object" and "unsafety from multiple inheritance".